### ISSUE TITLE
lint: Fail when using "New KR" IDs on objectives

### ISSUE THREAD

**Issue by gpetiot** [tarides/okra#274]
Similarly to #270 and #271, the linter should fail if the ID of an objective is: New KR / New OKR / New WI / No KR / No OKR / No WI, for the reports written from week 24.

cc @ganeshn-gh @wafaesara

**Comment by jmid** [comment]
What should engineers use instead when no objective has been created yet? `New Obj`? Lump the work under `Misc`?

**Comment by gpetiot** [comment]
I thought objectives were created when writing the proposals, so theoretically before the engineers start working on it.
Anyway reporting work on a non existing objective is an issue when aggregating the timesheets. @ganeshn-gh @kayceesrk

**Comment by jmid** [comment]
> I thought objectives were created when writing the proposals, so theoretically before the engineers start working on it.

In practice, they are not always opened yet.

Having some way to report still would be welcome, otherwise I expect some engineers just to file
- non-linting weeklies (worse)
- things under `Misc` (also worse, as it gets harder to separate time usage from other things in there)

**Comment by ganeshn-gh** [comment]
> In practice, they are not always opened yet.
> 
The proper practice is to create the `objectives` for all the proposals. 
> Having some way to report still would be welcome, otherwise I expect some engineers just to file
> 
> * non-linting weeklies (worse)
>
We recommend that the engineers check the linter. If the objective is not created, please inform TL/EMs.
> * things under `Misc` (also worse, as it gets harder to separate time usage from other things in there)
>
We are working to streamline the process. To make it a success, we have to communicate very clearly if there is any issue with the linter or objective.

**Comment by jmid** [comment]
> The proper practice is to create the `objectives` for all the proposals.

Sometimes part of the work is to write the proposal, which leaves us in a chicken-and-egg situation.

> We are working to streamline the process. To make it a success, we have to communicate very clearly if there is any issue with the linter or objective.

I'm trying to do that here. I don't see a hard error as streamlining.
As a TL, I would prefer to keep the old behaviour of `Some new kr (New KR)` getting merged with `Some new kr (#1234)` based on title automatically, if the former hasn't been opened in the database yet, rather than block or throw an error at the engineer.

**Comment by wafaesara** [comment]
@jmid  What prevents engineers from creating the objective in the objective repo to get the ID even if the proposal is still "First Draft"?  Is it a permission issue or a process issue? (only EMs can create them?)

**Comment by ganeshn-gh** [comment]
> > The proper practice is to create the `objectives` for all the proposals.
> 
> Sometimes part of the work is to write the proposal, which leaves us in a chicken-and-egg situation.
> 
I'm unable to understand the chicken-and-egg situation here.
If you are writing the proposal, you should also be able to create an objective. I can't see the roadblock here. 
> > We are working to streamline the process. To make it a success, we have to communicate very clearly if there is any issue with the linter or objective.
> 
> I'm trying to do that here. I don't see a hard error as streamlining. As a TL, I would prefer to keep the old behaviour of `Some new kr (New KR)` getting merged with `Some new kr (#1234)` based on title automatically, if the former hasn't been opened in the database yet, rather than block or throw an error at the engineer.
>
When `New KR` is assigned, my job becomes difficult. If the report had the correct objective, I wouldn't have to search for the objective details. Instead, I have to follow up with the engineer/TL or their EM. This consumes a lot of my time, and imagine if I have to do it for many reports.

**Comment by jmid** [comment]
> What prevents engineers from creating the objective in the objective repo to get the ID even if the proposal is still "First Draft"?

I don't think it is a good idea to encourage engineers to make changes to the database, when the same engineers are apparently sometimes misfiling their work.

> When `New KR` is assigned, my job becomes difficult. [...]

Shouldn't identifiers be resolved by running a report through `okra cat`? @rikusilvola

**Comment by gpetiot** [comment]
If you run okra cat before the objectives are created it won’t help. The weeklies are processed by Ganesh soon after they are submitted (1 week?), so the engineers have to create the objectives before that.

**Comment by jmid** [comment]
Weeklies are processed on a monthly basis, no? https://github.com/tarides/admin?tab=readme-ov-file#weekly-submission-deadlines-for-2024

**Comment by wafaesara** [comment]
> I don't think it is a good idea to encourage engineers to make changes to the database, when the same engineers are apparently sometimes misfiling their work.

Can we communicate that they ask either a TL, EM or me to create the objective before they submit their weekly reports? 

Having a 'New ID' (or new KR) makes it difficult for those who consume the weekly data, and replacing each 'New KR' afterward becomes too time-consuming and is not scalable. 

It has been agreed that time spent researching and writing a proposal should be logged against that proposal (if it is not accepted, it can be marked as dropped later). Therefore, creating the proposal or the objective early on should not be an issue.

**Comment by wafaesara** [comment]
> Weeklies are processed on a monthly basis, no? https://github.com/tarides/admin?tab=readme-ov-file#weekly-submission-deadlines-for-2024

It is processed manually. We cannot wait one month before we start processing all the weeklies from all engineers (it won't be feasible). So, the reports are processed on a rolling basis as soon as they are submitted

**Comment by ganeshn-gh** [comment]
When manually adding the data in the past(continuing for June), I ignored the linting output and entered the weekly data. I did not ping the engineers when there were issues in the report, so I added the data when the weeklies were added.

But right now, we are trying to automate the process of extracting the data from the weeklies using the caretaker. To ensure the caretaker runs well, we have to make sure linting passes. So, I have started nudging people if their weeklies fail linting.

As you said, @jmid, even I want to work on this process on a monthly basis.

**Comment by ganeshn-gh** [comment]
> Weeklies are processed on a monthly basis, no? https://github.com/tarides/admin?tab=readme-ov-file#weekly-submission-deadlines-for-2024

Also, not everyone submits their reports on or before the deadline. That makes it challenging to follow up, wait, and update the data repeatedly.
